Tasks/SonarQubePostTest/PRCA/PostComments-Module.psm1

```diff
@@ -181,39 +299,190 @@ function GetPullRequestId
+    param ([ValidateNotNull()][Array]$messages, [string][ValidateNotNullOrEmpty()]$messageSource)
     
-    # Limit the number of messages so as to not overload the PR with too many comments
-    $comments = $comments | Sort-Object Priority | Select-Object -first $PostCommentsModule_MaxIssuesToPost
-    Write-Host "Sorting comments and filtering before postng"
+    Write-Verbose "Fetching existing threads and comments..."
+    $existingThreads = FetchDiscussionThreads 
+    $existingComments = FetchDiscussionComments $existingThreads    
     
-    $comments | ForEach {Write-Verbose $_} 
+    $messages = FilterComments $messages $messageSource $existingThreads $existingComments
     
-    # TODO: check that the comments aren't already present before posting
+    $messages | ForEach {Write-Verbose $_} 
     
-    $newDiscussionThreads = CreateDiscussionThreads $comments
+    $newDiscussionThreads = CreateDiscussionThreads $messages $messageSource
     PostDiscussionThreads $newDiscussionThreads 
 }
 
+# region Filter Comments
+function FilterComments 
+{
+    param ([Array]$messages, 
+    [ValidateNotNullOrEmpty()][string]$messageSource, 
+    [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+    [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+    
+    Write-Verbose "Filtering comments before posting"
+    
+    $messages = FilterCommentsByPath $messages
+    $messages = FilterPreExistingComments $messages $messageSource $existingThreads $existingComments
+    $messages = FilterMessagesByNumber $messages
+    
+    return $messages
+}
+
+function FilterCommentsByPath
+{
+    param ([Array]$messages)
+    
+    $modifiedFilesInPr = GetModifiedFilesInPR
+    Write-Verbose "Files changed in this PR: $modifiedFilesInPr"
+    
+    $countBefore = $messages.Count
+    $messages = $messages | Where-Object {$modifiedFilesInPr.Contains($_.RelativePath)}
+    $commentsFiltered = $countBefore - $messages.Count 
+    
+    Write-Host "$commentsFiltered message(s) were filtered because they do not belong to files that were changed in this PR"
+    
+    return $messages
+}
+
+#
+# A message is said to be pre-existing if:
+# 1. a comment with the same content exists 
+# 2. and it belongs to the same file 
+# 3. and the thread was created by the same logic, i.e. the same commentSource 
+#
+# Remark: comments move arround so Line cannot be used. 
+#
+function FilterPreExistingComments
+{
+     param ([Array]$messages, 
+     [ValidateNotNullOrEmpty()][string]$messageSource, 
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+     
+     $sw = new-object "Diagnostics.Stopwatch"
```

Comment:
I am somewhat concerned about this logic so I added a stopwatch and will do more E2E testing to asses how slow this is. The algorithm is: 
foreach message to be posted ...
foreach thread that points to the same file ...
foreach comment that has the same content....

So quite intensive. The original implementation simply gives up if there are more than 1000 comments to add, but I did not add this limit yet.